File: 76-minimum-window-substring/76-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        m, n =len(s), len(t)
        count = dict()
        for char in t:
            if char in count:
                count[char] += 1
            else:
                count[char] = 1
        left, right = 0, 0
        def equals(l1, l2):
            if not l1:
                return False
            for char,num in l2.items():
                if char not in l1:
                    return False
                if char in l1 and l1[char] < num:
                    return False
            return True
        cur = dict()
        res = ''
        while(right < m and left <= right):
            while(right<m and not equals(cur,count)): 
                if s[right] in cur:
                    cur[s[right]] += 1
                else:
                    cur[s[right]] = 1
                right = right+1
            while(left < right and equals(cur,count)):
                if cur[s[left]] > 1:
                    cur[s[left]] = cur[s[left]] - 1
                else:
                    del cur[s[left]]
                if (res == '' or right-left < len(res)):
                    res = s[left:right]
                left = left + 1
        return res
            
            
            
        
        

File: 76-minimum-window-substring/76-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        // rolling hash
        HashMap<Character, Integer> windowMap = new HashMap<>();
        HashMap<Character, Integer> targetMap = new HashMap<>();
        int left = 0, right = 0;
        int N = s.length();
        int ans = Integer.MAX_VALUE;
        String res = "";
        constructMap(t, targetMap);
        while(right < N){
            char curChar = s.charAt(right);
            windowMap.computeIfAbsent(curChar, k->0);
            windowMap.put(curChar, (int) windowMap.get(curChar) + 1);
            right++;
            while(left < right && compareMap(targetMap, windowMap)){
                char leftChar = s.charAt(left);
                windowMap.put(leftChar, windowMap.get(leftChar) - 1);
                if(ans > right - left){
                ans = Math.min(ans, right - left);
                    res = s.substring(left, right);
                }
                left++;
            }
        }
        return res;
    }
    
    public void constructMap(String t, Map targetMap){
        for(int i = 0; i < t.length(); i++){
            char curChar = t.charAt(i);
            targetMap.computeIfAbsent(curChar, k -> 0);
            targetMap.put(curChar, (int) targetMap.get(curChar) + 1);   
        }
    }
    public boolean compareMap(Map<Character, Integer> targetMap, Map<Character, Integer> windowMap){
        for(Map.Entry<Character, Integer> entry : targetMap.entrySet()){
            char key = entry.getKey();
            int freq = entry.getValue();
            if(!windowMap.containsKey(key) || windowMap.get(key) < freq)return false;
        }
        return true;
    }

}